### PR TITLE
Charlesmchen/conversation view crash logging

### PR DESCRIPTION
**Pull request by charlesmchen-signal**
PTAL @michaelkirk

### CODE REVIEW

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 1]
```diff
@@ -4894,6 +4916,14 @@ - (void)reloadViewItems
         }
     }];
 
+    // Flag to ensure that we only increment once per launch.
+    static BOOL hasIncrementedDatabaseView = NO;
+    if (hasError && !hasIncrementedDatabaseView) {
+        DDLogWarn(@"%@ incrementing version of: %@", self.logTag, TSMessageDatabaseViewExtensionName);
+        [OWSPrimaryStorage incrementVersionOfDatabaseExtension:TSMessageDatabaseViewExtensionName];
+        hasIncrementedDatabaseView = YES;
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: This class of error indicates that we need to rebuild our database view.  I've confirmed through Riya that this resolves the issue.

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 2]
```diff
@@ -3378,14 +3363,21 @@ - (void)uiDatabaseDidUpdate:(NSNotification *)notification
...
-        [self resetContentAndLayout];
+        [self resetMappings];
         [self updateLastVisibleTimestamp];
         [self scrollToBottomAnimated:NO];
         return;
     }
 
     NSUInteger oldViewItemCount = self.viewItems.count;
-    [self reloadViewItems];
+    if (![self reloadViewItems]) {
```
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: This makes the view robust to this issue until we can rebuild the database view on next launch, which should address a number of our worst crashes.

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 3]
```diff
@@ -4800,6 +4792,32 @@ - (void)extendRangeToIncludeUnobservedItems
     self.previousLastTimestamp = nil;
 }
 
+- (void)createNewMessageMappings
+{
+    if (self.thread.uniqueId.length > 0) {
+        self.messageMappings = [[YapDatabaseViewMappings alloc] initWithGroups:@[ self.thread.uniqueId ]
+                                                                          view:TSMessageDatabaseViewExtensionName];
+    } else {
+        OWSFail(@"uniqueId unexpectedly empty for thread: %@", self.thread);
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: huh - is this ever known to happen?

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 4]
```diff
@@ -4800,6 +4792,32 @@ - (void)extendRangeToIncludeUnobservedItems
...
+    }
+
+    // Cells' appearance can depend on adjacent cells in both directions.
+    [self.messageMappings setCellDrawingDependencyOffsets:[NSSet setWithArray:@[
+        @(-1),
+        @(+1),
+    ]]
+                                                 forGroup:self.thread.uniqueId];
+
+    // We need to impose the range restrictions on the mappings immediately to avoid
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: seems like this comment applies to the code 3 lines down.

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 5]
```diff
@@ -455,29 +457,11 @@ - (void)configureForThread:(TSThread *)thread
...
-    if (thread.uniqueId.length > 0) {
-        self.messageMappings = [[YapDatabaseViewMappings alloc] initWithGroups:@[ thread.uniqueId ]
-                                                                          view:TSMessageDatabaseViewExtensionName];
-    } else {
-        OWSFail(@"uniqueId unexpectedly empty for thread: %@", thread);
-        self.messageMappings =
-            [[YapDatabaseViewMappings alloc] initWithGroups:@[] view:TSMessageDatabaseViewExtensionName];
-        return;
+    [self createNewMessageMappings];
+    if (![self reloadViewItems]) {
```
[the remark this answers is not in the document]
Comment (michaelkirk-signal) [reply Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: 👍

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m (hunk 3 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: There was an "extremely hard to repro" bug of this form 1+ years ago.  I haven't seen it in the last year, but I know I wasn't the only person who saw it then.

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m (hunk 4 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: Done.

**Pull request merged**